### ISSUE TITLE
Canonical Paths

### ISSUE THREAD

**Issue by jaheba** [awslabs/gluon-ts#1606]
In light of #1592, I came to think that we should have canonical paths for out entities.

Motivating example:

```python
from gluonts.mx import Trainer
from gluonts.mx.trainer import Trainer
from gluonts.mx.trainer._base import Trainer
```

All three versions are possible, but arguably the first one should be the “official” one.

This should be reflected in:

* the documentation, so we would document `gluonts.mx.Trainer`
* in serde, exporting an object of the class should use `gluonts.mx.Trainer` for the path to the class
* in imports, I think we are already covering this.

I think sphinx already makes use of re-exports through `__all__` to shorten the class-paths.

However, `serde.encode` currently does not.

The solution would be to rewrite the `__module__` of the objects to anchor them where we want it. It appears, we are already doing something like that, but the code I saw doesn’t work, since it checks `__module__` of the strings in `__all__` instead of the actual entities.

I’ve played around a bit, and would propose something like this:

```python
# mx/__init__.py
from .trainer import Trainer

re_export(Trainer)

__all__ = [“Trainer”, …]
```

`re_export` would set the `__module__` path accordingly.

**Comment by jaheba** [comment]
Code for `re_export`:

```python

import inspect


def re_rexport(obj):
    context = inspect.stack()[1]
    module = inspect.getmodule(context[0])

    obj_path = obj.__module__.split(".")
    module_path = module.__name__.split(".")

    assert obj_path[: len(module_path)] == module_path

    obj.__module__ = module.__name__

    return obj
```

**Comment by lostella** [comment]
I agree fully on using short imports in the documentation, tests and throughout the package. What would be the benefit of the serde story? Avoid breaking objects when a class is moved around?

**Comment by jaheba** [comment]
> I agree fully on using short imports in the documentation, tests and throughout the package. What would be the benefit of the serde story? Avoid breaking objects when a class is moved around?

Yes, compatibility and a bit of readability.

**Comment by jaheba** [comment]
After working more intensely with sphinx, we should make this a priority.

One thing sphinx does is to cross-link references.

For example, we document create a documentation entry for `gluonts.mx.Trainer`. However, the `trainer` argument of `DeepAR` is `gluonts.mx.trainer._base.Trainer` which means that sphinx does not create a cross-link. If we change the `__module__` of `Trainer` to be `gluonts.mx` it works.

Maybe we can create a convention where we prefix sub-modules that we don't want to expose with an `_`. That would signal us that the definitions of that module are re-exported at a higher level -- or don't intend to be exposed at all. I think we follow that convention already in some places, but not everywhere.
